### ISSUE TITLE
cursorA.equals(cursorB) should check _keyPath

### ISSUE THREAD

**Issue by Gozala** [facebook/immutable-js#370]
You can find more details about the issues here: https://github.com/omniscientjs/omniscient/issues/89

Here is a little illustrative summary:

``` js
let root = Cursor.from(Immutable.fromJS({items: [{x: 3}, {x: 2}, {x: 1}]}))
let x1 = root.cursor('items').cursor(0);

let root2 = root.updateIn(['items'], items => items.reverse());
let x2 = root2.cursor('items').cursor(2);

x1.equals(x2) // => true
x1._keyPath === x2._keyPath // => false
```

Now consider react components like this:

``` js
class Item extends React.Component {
  shouldComponentUpdate(props)  {
     return !this.props.cursor.equals(props.cursor);
  }
  render() {
    const {cursor} = this.props;
    return React.DOM.div({
      onClick() {
         cursor.update(...)
      }
    }, generateContent(cursor))
  }
});

class App Extends React.Component {
   render() {
     const children = this.props.cursor.cursor('items').map(item =>
        Item({ key: item.get('x'), cursor: item }));
      return React.DOM.div({}, children);
   }
}
```

If `App` is rendered with with `root` and then `root2` Item for `{x: 3}` will not re-render as equality check returns `true` even though it refers to different location in the structure. Also clicking it will end up sending changes to out of date `_keyPath`.

With react API this is actually not a big deal if one would not capture cursor in render, but in omniscient it is because cursor will be passed in into `render` itself.

Maybe `.equals` behavior is correct, but what would be a way to handle such case properly ?

**Comment by leebyron** [comment]
Interesting. I think you're correct here that if the key paths differ then the cursors can't be considered equal because they refer to different parts of the source structure.

However if a cursor is compared to a non-cursor, the key path should be inconsequential.

Thanks for the really detailed write-up about this.

**Comment by Gozala** [comment]
To be honest I can get behind over either point of views on this matter. Although I am afraid there are issues with ether option. Here are my further thoughts on this:

1. If `.equals` takes key path  into account it should probably also take `root` data structure into account as well (or do they ?). If so this will lead to another issue where observer of the `root` cursor will initiate another render that would force all of the subcomponent to re-render as the sub-cursors will be pointing to a diff `root`.

2. If `.equals` continues to behave as is there is no public API right now to see if key path is different and tweak `shouldComponentUpdate` accordingly.

I'm personally leaning towards keeping `.equals` as it is now as structures themselves are equal. To make an analogy structures can be equal regardless of where they came from:

```js
const m = Immutable.map({ x: x, y: {x: x} });
m.get('x').equals(m.getIn(['y', 'x']) // => true
```

As a matter of fact in terms of react / omniscient re-rendering a component when a key path has changed is not ideal, ideal solution would be to update previous equal cursor to point to a new path (not sure if that still could lead to errors with a deep structures like when parent components cursor will be re-rooted while it's children will end up having old root because render short circuited).

Maybe just exposing key path in the public API is enough ?

I was also wondering if you have considered alternative cursor API, like the one used by Om. My understanding of it is that cursors get synchronised on updates there for same component structure won't lead to any errors as render will short circuit while cursors will still get updated (similar to how `this.props` get updated regardless of what `shouldComponentUpdate` returned).

**Comment by wmertens** [comment]
@Gozala am I understanding correctly that your onClick() handler changes the cursor directly? That breaks top-down flow.
The way I use cursors is by providing an onChange handler that gives the new root object value to the store and returns undefined so the cursor keeps pointing to the previous version of the root object.
Then the store-triggered redraw handles everything as it should, since the cursor content will differ in the new props.

**Comment by Gozala** [comment]
> @Gozala am I understanding correctly that your onClick() handler changes the cursor directly? That breaks top-down flow.

Hi @wmertens thanks for your input. I am afraid I do not agree that changing cursors directly from the event handlers breaks top down flow. Although depends what do you mean by changing directly. The purpose of cursors is pretty much what the purpose of stores is (at least in our use of them). Root component is passed cursor for the whole app state which is observed for changes (note that changes to cursors don't actually mutate anything just trigger observer with a whole new data structure with changes applied) once observer is invoked root component is re-rendered with a new cursor created from the updated state. So it is the same top down flow and same idea as stores.

> The way I use cursors is by providing an onChange handler that gives the new root object value to the store and returns undefined so the cursor keeps pointing to the previous version of the root object.

It sounds pretty much what happens here as well, cursor indeed will still point to the old root object.

> Then the store-triggered redraw handles everything as it should, since the cursor content will differ in the new props.

This is where I think you are misunderstanding a problem. In our case simplified `shouldComponentUpdate` looks like this:

```js
{
  shouldComponentUpdate(newProps) {
     return !this.props.cursor.equals(newProps.cursor)
  }
}
```

Now if some user interaction just caused to reverse `items` on a next render each `item` from the list will still point to the same data structure and cursor will be equal although their `._keyPath` will be different. Which would lead to an issue that is that `shouldComponentUpdate` on each item will return `false` there for `render` will not execute and there for when `onClick` is invoked enclosed `cursor` on each item will point to old root and old `._keyPath`. Note: That this would not be an issue if `onClick` handler referred to `this.props.cursor` instead of enclosed `cursor` as `props` would have updated even though `render` was not triggered.

My guess reason you do not run into this issues is either because your `shouldComponentUpdate` does not return `false` even if state has not changed (maybe because you pass new store instance or something else) or because your `onClick` invokes something on the mutable object, like store who's target cursor was updated or `this.props` that are mutated for you by react.

I hope this explains it.

**Comment by wmertens** [comment]
@Gozala The way it works for me is because the list itself is also rendered with a cursor, and that will point to a different object even though the `item`s themselves do not change.

Also, I only interface with the store in "container" elements that create cursors and render "pure" elements.

Here's my flow (I use Fluxy):
`onclick() -> cursor.update() -> _onChange(newRoot) -> store saves newRoot -> notifies watchers -> redraws`

`_onChange` returns undefined so the cursors remain stale. The store watchers are only set on container elements that wrap the pure elements. They simply call `this.forceUpdate()` which causes new cursors to be created from the store key they're watching.

So in fact it's `click -> new root data -> create new cursors` and then the wrapped pure elements do conditional rendering. I do check with Perf to see if there were wasted updates so I know it works as it should.

**Comment by Gozala** [comment]
> @Gozala The way it works for me is because the list itself is also rendered with a cursor, and that will point to a different object even though the items themselves do not change.

In our case everything is cursor from top to bottom as well (as you can see in the description `App` component is using `this.props.cursor.cursor('items')` to get a cursor to the list. But since each `item` is handler with a `Item` component each of them will run `shouldComponentUpdate` and cursor equality checks will return `true` even though they belong to different structure and in a different path. Do you see what I mean ?

> Also, I only interface with the store in "container" elements that create cursors and render "pure" elements.

> Here's my flow (I use Fluxy):
onclick() -> cursor.update() -> _onChange(newRoot) -> store saves newRoot -> notifies watchers -> redraws

Are you doing `cursor.update()` in `onClick` or call method on the store ? If later that would absolutely explain why you don't have this issue, because your store will update to track a new cursor so it's method will end up operating on up to date cursor regardless if render was called or not. If you do `cursor.update()` from a `onClick` handler & indeed don't trigger render on each item after reorder then I'm really curios to see the actual code as I don't see how that could be a case. If that a case maybe you could modify this example to illustrate your approach http://jsbin.com/yunomateha/1/edit?js,output otherwise it's hard to follow what's going on. (P.S. that example uses cursors wrapper from omniscient library as I wrote it for https://github.com/omniscientjs/omniscient/issues/89 but same happens on plain Immutable.js Cursors and plain react).

> _onChange returns undefined so the cursors remain stale. The store watchers are only set on container elements that wrap the pure elements. They simply call this.forceUpdate() which causes new cursors to be created from the store key they're watching.

> So in fact it's click -> new root data -> create new cursors and then the wrapped pure elements do conditional rendering. I do check with Perf to see if there were wasted updates so I know it works as it should.

I'm sorry at this point it's really hard for me to understand how things work without seeing actual code. Wait does not `.forceUpdate()` is going to call the `render` though ? Also this issue has nothing to do with weather new cursors are created or not, the issue is that two absolutely different cursors with different roots (that may not even have much in common) pointing to the same data structure (in totally different places) are considered equal & there for `shouldComponentUpdate` will return `false`.

**Comment by wmertens** [comment]
I'll take a look at the jsbin later, but I think your problem is due to
this:

> Also this issue has nothing to do with weather new cursors are created or
not, the issue is that two absolutely different cursors with different
roots (that may not even have much in common) pointing to the same data
structure (in totally different places) are considered equal & there for
shouldComponentUpdate will return false.

When a pure component gets the same props, it should render the same way.
In this case, it gets a cursor which doesn't expose the path, so the props
are the same and the element should render the same. I think what's going
on is that your list of items doesn't re-render when it should. If it did
re-render, it would have created the elements in a different order with the
same keys and React would have moved the DOM elements.

**Comment by Gozala** [comment]
> When a pure component gets the same props, it should render the same way.
> In this case, it gets a cursor which doesn't expose the path, so the props
> are the same and the element should render the same. I think what's going
> on is that your list of items doesn't re-render when it should. If it did
> re-render, it would have created the elements in a different order with the
> same keys and React would have moved the DOM elements.

List itself does get  reordered, but react does not executes render methods of an Item components and reason is exactly what this bug say if `a1` was moved to a different index and become `a5` `a1.equal(a5)`  will be `true` so react will not run `render` and just use previously generated VDOM.

**Comment by wmertens** [comment]
I don't see how that is a bug - I think that's exactly how it should work?

Oh - are you using the array index as the Item key? You should use a
per-item unique key so that React knows what goes where. Otherwise React
will just see "item with key 1 - still the same, great, nothing to do; item
with key 5 - still the same, great, nothing to do".

**Comment by Gozala** [comment]
> I don't see how that is a bug - I think that's exactly how it should work?

This is not a bug indeed, but it has an unfortunate side effect, which is since render will not run event listeners from previous run will hold on to a previous cursor that has a different ._keyPath so if the will fire they will end up updating wrong state.



> Oh - are you using the array index as the Item key? You should use a
> per-item unique key so that React knows what goes where. Otherwise React
> will just see "item with key 1 - still the same, great, nothing to do; item
> with key 5 - still the same, great, nothing to do".

Please read the code I provided with the bug, I feel I keep explaining what is already illustrated in the code snippet provided.

**Comment by wmertens** [comment]
So the way I do it is that cursor.update tells the store to update but **does not** change the cursor root data value. Then the store triggers a re-render where needed (I register store listeners on wrapper elements that talk to the store).

Your List would be a different cursor and it would cause DOM element moves.

```es6
class Item extends React.Component {
  shouldComponentUpdate(props)  {
     return !this.props.cursor.equals(props.cursor);
  }
  render() {
    const {cursor} = this.props;
    return React.DOM.div({
      onClick() {
         cursor.update(...)
      }
    }, generateContent(cursor))
  }
});

class App Extends React.Component {
   render() {
     const children = this.props.cursor.cursor('items').map(item =>
        Item({ key: item.get('x'), cursor: item }));
      return React.DOM.div({}, children);
   }
}

class AppStoreWrapper Extends React.Component {
  // Somehow listen to store changes induced by cursor updates
  // Cursor onChange callback calls store directly or launches action and returns undefined
  // When the store is done updating, it should notify its listeners (maybe just render the whole app)
  // and they will get new cursors that will be different from the originals when the data needs updating.
 render() {
  cursor = store.getCursor()
  props = extend({cursor: cursor},this.props)
  return React.createElement(props, App);
 }
}```

**Comment by wmertens** [comment]
@Gozala I just found an annoying issue with my approach, I'm not sure if it's what you're seeing: Since cursors only compare their referenced value, they can nicely avoid repainting whole blocks, but when you make use of cursor.update() the way I describe, it will use stale data to make the change on itself.

In other words, you have an app that paints `{a: {b: [c, d]}}` into a section using `a.b.0` and another using `a.b.1`, and then the user operates on `c`, resulting in `{a2: {b2: [c2, d]}}` for the app state, but since `d` didn't change, it will still refer to the original data structure. When the user then changes `d`, the `onChange()` handler gets the old `a` with a change to `d`, and the change to `c` is lost.

To fix this, the `onChange()` handler should get the partial data...

Is this related to your issue? I think not right?

**Comment by Gozala** [comment]
> @Gozala I just found an annoying issue with my approach, I'm not sure if it's what you're seeing: Since cursors only compare their referenced value, they can nicely avoid repainting whole blocks, but when you make use of cursor.update() the way I describe, it will use stale data to make the change on itself.

Yes this is another example of this issue, although this one is less problematic scenario than the one I described.



> To fix this, the onChange() handler should get the partial data...

There are different workaround in different scenarios and neither issues nor solutions are obvious. Actual code base I run in those issues is complex enough that workarounding this solution was not a viable option as we would run into different variations of it over and over.

That is why I originally proposed to modify cursor equality check so it would also compare `.keyPaths` that way render will be triggered even if data structure has not changed but only it's location did. Never the less that isn't ideal options either as that would cause unnecessary render calls just to update cursors so event handlers will have up to date reference.

At this point I think most viable solution going with cursors is to do what https://github.com/omcljs/om does which always keeps cursors in sync with changes, although I'm not in a favour of that approach either as at that point cursors don't seem any better than mutable state.

In the end I'm planning on moving from cursors to an approach inspired by http://elm-lang.org and illustrated in this example http://jsbin.com/hogulenuhu/5/edit?js,console,output

**Comment by wmertens** [comment]
I still like cursors as an immutable proxy for a store object, so I changed them so they never change and only pass the changed sub-object and path to the store. The store can then decide what to do with it.

This fixes my issues.

I agree that the Om approach doesn't make sense in a React context.

**Comment by Gozala** [comment]
> I agree that the Om approach doesn't make sense in a React context.

@wmertens I think it might be my imperfect English but I feel like we talk past each other as you misunderstand comments I'm making.

I did not implied Om approach does no make sense in React context. In fact I think only reasonable options of going with cursors is to follow Om's approach.

What I tried to say was that **I personally** came to realise that cursors that get synced automatically is **not much better** than a mutable API exposed by react. 


> I still like cursors as an immutable proxy for a store object, so I changed them so they never change and only pass the changed sub-object and path to the store. The store can then decide what to do with it.
> 
> In other words, cursors are immutable too.
> 
> This fixes my issues.
> 

I am afraid I don't fully understand your solution or how it solves this problem without causing unnecessary re-render. In fact it seems odd to me that you are using both cursors and stores as they are tools that take different approach to solve the same problem - provide a single source of truth.

Don't get me wrong I'm not implying there is anything wrong with your use, it's just does not seems to be a good fit for my use. And even if such a combination of tools avoids described issue I still think issue is present and is worth thinking about a fix that would not require additional abstractions. Again I tend to lean towards the Om like cursors as that seems to work and be free of this problem.

**Comment by wmertens** [comment]
The way I understand it from you, Om cursors are updated immediately all over the application. If that is correct, that means that when you are comparing cursors to see if an element needs updating by React, they are actually the same.
So actually that probably can't work, unless you deep copy the cursor state at render() time.

I use cursors as the replacement for a (Store, object, path) triplet and I keep them immutable. When you call one of the change methods, the store gets notified about the change and then propagates the change to the elements that watch the store. This propagation generates new cursors which can compare against the old ones and thus prevent unnecessary redraws. 

I do not have an issue with elements moving in a list, that simply means the list gets re-rendered and DOM elements moved to their correct place.

**Comment by facebook-github-bot-6** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by leebyron** [comment]
Closing this aging issue, especially since contrib/ is not a primary focus. If the changes discussed here are still relevant to someone then a PR would be a nice reboot for this concern
